Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,486 @@
+    std::vector<amrex::ParticleReal> currY(numParticleOnTile, 0);
+    std::vector<amrex::ParticleReal> currZ(numParticleOnTile, 0);
+
+    for (auto i=0; i<numParticleOnTile; i++) {
+      currX[i] = aos[i].m_rdata.pos[0];
+      currY[i] = aos[i].m_rdata.pos[1];
+      currZ[i] = aos[i].m_rdata.pos[2];
+    }
+
+    currSpecies["position"]["x"].storeChunk(currX, {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+    currSpecies["position"]["y"].storeChunk(currY, {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+    currSpecies["position"]["z"].storeChunk(currZ, {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+
+    SaveRealProperty(pti,
+             currSpecies,
+             offset, numParticles,
+             write_real_comp, real_comp_names);
+#endif
+
+    offset += numParticleOnTile;
+      }
+    }
+}
+void
+WarpXOpenPMDParticle::SetupRealProperties(openPMD::ParticleSpecies& currSpecies,
+                      const amrex::Vector<int>& write_real_comp,
+                      const amrex::Vector<std::string>& real_comp_names,
+                      unsigned long long np) const
+{
+  auto particlesLineup = openPMD::Dataset(openPMD::determineDatatype<amrex::ParticleReal>(), {np});
+
+  //
+  // the beam/input3d showed write_real_comp.size() = 16 while only 10 real comp names
+  // so using the min to be safe.
+  //
+  auto counter = std::min(write_real_comp.size(), real_comp_names.size());
+  for (int i = 0; i < counter; ++i)
+    if (write_real_comp[i]) {
+      auto& particleVar = currSpecies[real_comp_names[i]];
+      auto& particleVarComp = particleVar[openPMD::RecordComponent::SCALAR];
+      particleVarComp.resetDataset(particlesLineup);
+    }
+}
+
+void
+WarpXOpenPMDParticle::SaveRealProperty(WarpXParIter& pti,
+                       openPMD::ParticleSpecies& currSpecies,
+                       unsigned long offset,
+                       unsigned long long numParticles,
+                       const amrex::Vector<int>& write_real_comp,
+                       const amrex::Vector<std::string>& real_comp_names) const
+
+{
+  int numOutputReal = 0;
+  int totalRealAttrs = m_NumAoSRealAttributes + m_NumSoARealAttributes;
+
+  for (int i = 0; i < totalRealAttrs; ++i)
+    if (write_real_comp[i])
+      ++numOutputReal;
+
+  auto numParticleOnTile = pti.numParticles();
+  const auto& aos = pti.GetArrayOfStructs();  // size =  numParticlesOnTile
+  const auto& soa = pti.GetStructOfArrays();
+
+  // properties are saved seperately
+  {
+    for (auto idx=0; idx<m_NumAoSRealAttributes; idx++) {
+      if (write_real_comp[idx]) {
+    auto& currVar = currSpecies[real_comp_names[idx]][openPMD::RecordComponent::SCALAR];
+    typename amrex::ParticleReal *d =
+      static_cast<typename amrex::ParticleReal*> (malloc(sizeof(typename amrex::ParticleReal) *  numParticleOnTile));
+
+    for (auto kk=0; kk<numParticleOnTile; kk++)
+      d[kk] = aos[kk].m_rdata.arr[AMREX_SPACEDIM+idx];
```

Comment:
I think we might be able to just repeatedly call `storeChunk` with the `openPMD::shareRaw` helper here, instead of copying all the data into a large, contiguous array first.

Reply:
I am not shareRaw applies  here.

Reply:
Oh I was mistaken: we reconstruct an SoA from the AoS AMReX particle attributes.